### PR TITLE
Add FNet text generation example

### PR DESCRIPTION
**Pull request by DarshanDeshpande**
This is a seq2seq example that uses the FNet encoder and the Cornell Movie Dialog Corpus.
Paper: https://arxiv.org/abs/2105.03824v3

@fchollet Let me know what you think or if there is something I should change :)

### CODE REVIEW

File: examples/nlp/text_generation_fnet.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+    return inputs, outputs
+
+
+questions, answers = load_conversations()
+
+"""
+### Creating and fitting the tokenizer
+"""
+
+tokenizer = text.Tokenizer(
```
Comment (fchollet) [inline examples/nlp/text_generation_fnet.py:99]: Please don't use the `preprocessingt.text` module, which is legacy. Instead, use a `TextVectorization` layer and a `tf.data.Dataset`. As you will find, it is actually simpler as well as more performant.

Basically:

1. Create a `Dataset` that yields string tensors of questions and answers
2. Map a `TextVectorization` layer into it (see https://keras.io/examples/nlp/text_classification_from_scratch/#two-options-to-vectorize-the-data)

File: examples/nlp/text_generation_fnet.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+## Creating the Decoder
+The decoder architecture remains the same as the one proposed by (Vaswani et al., 2017)
+in the original transformer architecture, consisting of an embedding, positional
+encoding, two masked multiheaded attention layers and finally the dense output layers
+
+
+"""
+
+
+class PositionalEmbedding(layers.Layer):
```
Comment (fchollet) [inline examples/nlp/text_generation_fnet.py:202]: If PositionalEmbedding and TransformerDecoder are taken from the Deep Learning with Python implementation (https://keras.io/examples/nlp/neural_machine_translation_with_transformer/), you should mention that.

File: examples/nlp/text_generation_fnet.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+The outputs of the FFT layer are complex in nature. To avoid processing complex layers,
+only the real part(magnitude) is extracted.
+
+The dense layers that follow the Fourier transformation act as convolutions applied on
+the frequency domain.
+"""
+
+
+class TransformerEncoder(layers.Layer):
```
Comment (fchollet) [inline examples/nlp/text_generation_fnet.py:170]: Use a name that reflects the fact that this is a FNet TransformerEncoder.

File: examples/nlp/text_generation_fnet.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+    decoder_outputs = layers.Dense(VOCAB_SIZE, activation="softmax")(x)
+    decoder = keras.Model(
+        [decoder_inputs, encoded_seq_inputs], decoder_outputs, name="outputs"
+    )
+
+    decoder_outputs = decoder([decoder_inputs, encoder_outputs])
+    transformer = keras.Model(
+        [encoder_inputs, decoder_inputs], decoder_outputs, name="transformer"
+    )
+    return transformer
```
Comment (fchollet) [inline examples/nlp/text_generation_fnet.py:309]: Again, naming it `fnet` would be more specific.

File: examples/nlp/text_generation_fnet.py [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+)
+
+"""
+The default `epochs` parameter is set to a single epoch but the model will take around
+20-30 epochs of training to start outputting comprehensible sentences. Although accuracy
+is not a good measure for this task, we will use it just to get a hint of the improvement
+of the network.
+"""
+
+transformer.fit(dataset, epochs=1)
```
Comment (fchollet) [inline examples/nlp/text_generation_fnet.py:328]: Make sure to monitor performance on a validation set during training. See our guide: https://github.com/keras-team/keras-io/blob/master/contributor_guide.md

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: examples/nlp/text_generation_fnet.py (hunk 1 above)
Comment (DarshanDeshpande) [reply examples/nlp/text_generation_fnet.py]: I have made the changes. Please have a look

File: examples/nlp/text_generation_fnet.py (hunk 2 above)
Comment (DarshanDeshpande) [reply examples/nlp/text_generation_fnet.py:200]: I had cited the machine translation example at the bottom of the notebook but since you have mentioned it here, I have added it here as well.

File: examples/nlp/text_generation_fnet.py (hunk 3 above)
Comment (DarshanDeshpande) [reply examples/nlp/text_generation_fnet.py]: Changed to `FNetEncoder` and `FNetDecoder`

File: examples/nlp/text_generation_fnet.py (hunk 4 above)
Comment (DarshanDeshpande) [reply examples/nlp/text_generation_fnet.py]: Changed

File: examples/nlp/text_generation_fnet.py (hunk 5 above)
Comment (DarshanDeshpande) [reply examples/nlp/text_generation_fnet.py]: Used a validation dataset. Please check.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! The code looks great, just a couple minor nits! I'm also going to do a final round of copyedits once the code is ready.

File: examples/nlp/text_generation_fnet.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+
+
+vectorizer = layers.TextVectorization(
+    VOCAB_SIZE,
+    standardize=preprocess_text,
+    output_mode="int",
+    output_sequence_length=MAX_LENGTH,
+)
+
+vectorizer.adapt(tf.data.Dataset.from_tensor_slices((questions + answers)).batch(128))
```
[the remark this answers is not in the document]
Comment (DarshanDeshpande) [reply examples/nlp/text_generation_fnet.py:125]: Done. Please have a look

**Review by fchollet — APPROVED** [review]
Awesome, thanks for the update! I've pushed some copyedits. Everything looks great. Please pull the copyedits and add the generated files 👍

**Comment by DarshanDeshpande** [comment]
@fchollet I have added the generated files. Please have a look.

**Review by fchollet — APPROVED** [review]
Thanks for the great contribution! 👍
